scripts/proof_bot.py

```diff
@@ -0,0 +1,165 @@
+#!/usr/bin/python3
+
+from urllib.parse import urljoin
+
+import click
+import requests
+import rlp
+from ape import project
+from ape.api import AccountAPI
+from ape.cli import ConnectedProviderCommand, account_option
+from ape.contracts import ContractInstance
+from ape.exceptions import ContractLogicError
+from ape.logging import logger
+from eth_typing import HexStr
+from eth_utils import to_bytes, to_int
+
+EVENT_SIGNATURE = "0x8c5261668696ce22758910d05bab8f186d6eb247ceac2af2e82c7dc17669b036"
+EXIT_ALREADY_PROCESSED_ERROR = "EXIT_ALREADY_PROCESSED"
+
+
+def hex_to_bytes(data: str) -> bytes:
+    return to_bytes(hexstr=HexStr(data))
+
+
+def get_polygon_last_block_number(
+    account: AccountAPI, fx_base_channel_root_tunnel: ContractInstance
+) -> int:
+    """
+    Search in tx history for the last `receiveMessage` tx,
+    extracts block number (from Polygon) using that data
+    """
+
+    last_blocknumber = 0
+    for tx in account.history:
+        if tx.method_called and tx.method_called.name == "receiveMessage":
+            last_proof_data = hex_to_bytes(tx.transaction.model_dump()["data"])
+            last_proof = fx_base_channel_root_tunnel.decode_input(last_proof_data)[1][
+                "inputData"
+            ]
+            decoded = rlp.decode(last_proof)
+            blocknumber = to_int(decoded[2])
+            if blocknumber > last_blocknumber:
+                last_blocknumber = blocknumber
+
+    return last_blocknumber
+
+
+def get_message_sent_events(graphql_endpoint: str, last_blocknumber: int) -> list[dict]:
+    """
+    Queries GraphQL endpoint to retreive all new `MessageSent` events
+    on Polygon network
+    """
+
+    gql = (
+        """
+    query AllMessagesSent {
+    messageSents(where: {blockNumber_gte: """
+        + str(last_blocknumber)
+        + """}, orderBy: blockNumber) {
+        transactionHash
+    }
+    }
+    """
+    )
+
+    s = requests.session()
+    s.headers = {"Accept": "application/json", "Content-Type": "application/json"}
+
+    response = s.post(graphql_endpoint, json={"query": gql})
+
+    data = response.json()
+    messages = data["data"]["messageSents"]
+    return messages
+
+
+def push_proof(
+    account: AccountAPI, fx_base_channel_root_tunnel: ContractInstance, proof: bytes
+) -> bool:
+    """Sends `receiveMessage` tx with the provided proof"""
+
+    try:
+        fx_base_channel_root_tunnel.receiveMessage(proof, sender=account)
+        return True
+    except ContractLogicError as e:
+        if e.message != EXIT_ALREADY_PROCESSED_ERROR:
+            raise e
+        logger.info("Transaction was processed")
```

Comment:
```suggestion
        logger.info("Transaction already processed")
```
WDYT? less confusing in my opinion.